File: linked-list/linkedlist_from_file.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


# insert at end
def insert_at_end(head, value):
    new_node = Node(value)

    if head == None:
        return new_node
    current = head
    while current.next:
        current = current.next
    current.next = new_node
    return head


# Print the list
def print_list(head):
    current = head
    while current:
        print(f"{current.data}->", end="")
        current = current.next
    print("NULL")


# main function, as deleting is not necessary in python
if __name__ == "__main__":
    head = None

    try:
        with open("random_numbers.txt", "r") as file:
            # read the entire file content
            content = file.read()

            # Split the content by whitespace and convert to integers
            numbers = [int(num) for num in content.split()]

            # insert each number into the linkedlist
            for number in numbers:
                head = insert_at_end(head, number)
    except FileNotFoundError:
        print("Error: Ye file nahi mili")

    except ValueError:
        print("Error: Isme galat format ki values hai")

    print("Linked List:", end=" ")
    print_list(head)

File: linked-list/linkedlistFromFile.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure of the node
typedef struct Node {
  int data;
  struct Node* next;
} Node;

// Create a new node
Node* createNode(int value){
  Node* newNode = (Node*) malloc(sizeof(Node));
  if(newNode == NULL){
    printf("Memory allocation failed\n");
    exit(1);
  }
  newNode->data = value;
  newNode->next = NULL;
  return newNode;
}

// Insert a node at the end
void insertAtEnd(Node** headRef, int value){
  Node* newNode = createNode(value);
  if(*headRef == NULL){
    *headRef = newNode;
    return;
  }
  
  Node* current = *headRef;
  while(current->next != NULL){
    current = current->next;
  }
  current->next = newNode;
}

// Print the list
void printList(Node* head) {
    Node* current = head;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

// Free the list
void freeList(Node* head) {
    Node* current = head;
    while (current != NULL) {
        Node* temp = current;
        current = current->next;
        free(temp);
    }
}

int main(){
  Node* head = NULL;

  // opens the file in read mode
  FILE* file = fopen("random_numbers.txt", "r");
  int value;

  // read integers from file until EOF
  while(fscanf(file, "%d", &value) == 1){
    insertAtEnd(&head, value);
  }

  fclose(file);

  printf("Linked list from the file: ");
  printList(head);

  freeList(head);
  return 0;
}
